### ISSUE TITLE
some collections evoke errors when passed as parameter to a component

### ISSUE THREAD

**Issue by alewolf** [livewire/livewire#987]
### Description

In my `home.blade.php` I use a Livewire directive and need to pass two parameters like so:

`@livewire('price-table', ['table' => $table, 'local_country' => $local_country]) `

Both parameters are collections. 

When I use a drop down in my Livewire component on the front end I get the following error: 

```
Livewire\Exceptions\CorruptComponentPayloadException
Livewire encountered corrupt data when trying to hydrate the [price-table] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.
```

If I only pass `'local_country' => $local_country` as a parameter all works fine. 

If I only pass `'table' => $table` I get the error. 

When I create the table collection within the Livewire component itself, then all works fine too. 
But, since I need the `$table` in several parts of my page, I need to generate it in a controller and then pass it down to Livewire and other blade components. 

The only thing that is structurally different between $table and $local_country is that $table also contains relations. 

The problem only starts when I assign the table to a public variable with the mount method.

This evokes the error: 

```
public 	$table_y;
public 	$local_country_x;

public function mount($table, $local_country)
{
	$this->table_y = $table;
	$this->local_country_x = $local_country;
}
```

This works fine:

```
public 	$table_y;
public 	$local_country_x;

public function mount($table, $local_country)
{
	// $this->table_y = $table;
	$this->local_country_x = $local_country;
}
```

**Context**
- Livewire version: 1.1.0
- Laravel version: 7.11.0
- Browser: Chrome, Safari

**Comment by JayBizzle** [comment]
Is `$table` a base `Collection` or an `Eloquent` collection?

**Comment by alewolf** [comment]
`$table` is an `Illuminate\Database\Eloquent\Collection`

`$local_country` is a model with one item: `App\Country`

**Comment by alewolf** [comment]
I closed in to issue a bit more. 

`$table` is a collection of models with prices per item. 

In order to display the `$table` in a specific order on the first render I order it by one of the price fields that is present in each model like so: 

`$table = $this->prices->sortBy('ticket_price_for_adult');`

As soon as I remove the `sortBy('ticket_price_for_adult')` the error vanishes and LiveWire works as expected !

This works:

`$table = $this->prices;`

This confuses me, as I have no idea how in the world the sortBy() could 'damage' the collection and lets LiveWire run into problems. One guess is that there is some checksum in the collection that gets corrupted when re-ordering. If so, that might be a Laravel bug? Ideas?

**Comment by alewolf** [comment]
I've done some more testing and have narrowed down that the problem is created when passing down the `$table` through the `mount()` method. 

When passing down the `$table` **unsorted** to the mount method all works fine. 

But, when I sort the `$table` before passing it down to `mount()`, or when I pass it unsorted and then sort it within `mount()`,  then the problem appears. 

On the other hand, when I entirely create the $table within the LiveWire class, then sort and render it, all works fine. The difference here is, that `mount()` doesn't touch my `$table`.

I dumped both sorted $table collections, the one passed down through mount, and the one created within the LiveWire class, and visually compared them. The only difference that I spotted are the numbers that appear next to the collection names:

`Illuminate\Database\Eloquent\Collection {#15`

`Illuminate\Database\Eloquent\Collection {#411`
  
Everything else is visually **exactly** the same. 

![image](https://user-images.githubusercontent.com/7146794/82525565-f7201880-9b31-11ea-8cd2-b62e176f993a.png)

![image](https://user-images.githubusercontent.com/7146794/82525616-16b74100-9b32-11ea-8671-38db86c3c0fc.png)

**Comment by JayBizzle** [comment]
Yeah, we had this problem a couple of weeks ago. Never figure it out in the end, just worked around it

**Comment by alewolf** [comment]
I've managed to get a step further. 

Passing down the `$table` unsorted through `mount()` and then sort it with `sortBy()` within `render()` before calling `return view()` provokes the problem. So actually, I don't think `mount()` does any damage, but `render()` does.

```
public function render()
{
	$this->table = $this->table->sortBy('total_cost');

	return view('livewire.price-table', [
    	'table_paginated' => $this->table->paginate($this->perPage),
    ]);
}
```

But, when I revert the order of the `Eloquent Collection` back to the order of the keys, using the perPage drop down in the front end starts to work without throwing an error.

 ```
public function render()
{
	$this->table = $this->table->sortBy('total_cost');
	$this->table = $this->table->sortBy('total_cost')->sortKeys();

	return view('livewire.price-table', [
    	'table_paginated' => $this->table->paginate($this->perPage),
    ]);
}
```

So, using `values()` to reindex the keys after sorting the `Eloquent Collections` works as a workaround. 

 ```
public function render()
{
	$this->table = $this->table->sortBy('total_cost');
	$this->table = $this->table->sortBy('total_cost')->values();

	return view('livewire.price-table', [
    	'table_paginated' => $this->table->paginate($this->perPage),
    ]);
}
```

This means `render()` is not able to process `Eloquent Collections` properly when the keys are not sorted ascending. 

I will keep this open, since I believe this should be addressed in the code. I hope the outlined details will help some developer to find the underlying issue to fix it. 

Also I have another issue on top of the one discussed here which manifests in a different way which is why I will post this in another bug report. 

I will try to dive into the code myself, but I am too much of a novice to be able to find and solve the problem myself quickly, especially because there is a lot going on between backend / frontend / PHP / JavaScript. 

Despite having run into LiveWire limitations on the first day I started using it, I find LiveWire an amazing attempt to get the concept of Vue baked into Laravel ! I love it!

**Comment by JayBizzle** [comment]
Yeah, this now rings a bell. We were having issues after sorting a the collection. I'm sure the issue that we had was that the collection keys became strings after the sort, bit I haven't got the code in front of me to check at the moment.

**Comment by alewolf** [comment]
`sortKeys()` works but `sortKeysDesc()` doesn't, which means the order has some relevance too.

**Comment by mblarsen** [comment]
Usually when sorting anything in PHP I call either `array_keys` for arrays or `->values()` for collections to avoid this sort of problems.

I'm not sure etiher of these would apply to your case. E.g. `$table->priecs->sortBy(..)->values()`

**Comment by tylerwiegand** [comment]
i spent an hour on this and resolved it by passing an array to the view instead of a collection.

i had a related issue that switched Collection(Collection()) into Collection(array()) but only after a refresh, initial load was fine. i ended up passing parts of the collection to the view and assembling them in the view <?php tags ?>, which worked. very odd.
